src/libraries/System.Linq/src/System/Linq/Average.cs

```diff
@@ -36,6 +37,38 @@ public static double Average(this IEnumerable<int> source)
             }
         }
 
+        private static double Average(ReadOnlySpan<int> span)
+        {
+            if (span.IsEmpty)
+            {
+                ThrowHelper.ThrowNoElementsException();
+            }
+
+            long sum = 0;
+            int i = 0;
+
+            if (Vector.IsHardwareAccelerated && span.Length >= Vector<int>.Count)
+            {
+                Vector<long> sums = default;
+                do
+                {
+                    Vector.Widen(new Vector<int>(span.Slice(i)), out Vector<long> low, out Vector<long> high);
```

Comment:
Are the bounds checks actually getting elided here for `new Vector<int>(span.Slice(i))`? This seems like its going to be doing a bunch of extra work each iteration

Reply:
I don't believe so.  But I also didn't want to introduce unsafe code.

Reply:
Some other places in the runtime use a helper built around `Unsafe.ReadUnaligned` and then the xplat helpers now have `Vector128.LoadUnsafe(ref T, nuint index)`.

There's never really going to be a "safe" way to do this unless the JIT gets special support, however. You functionally have some `T` and want to read 2-32 of that `T` (depending on what the actual type is). So the best you'll generally get is doing the right checks up front and potentially adding some `asserts`. Otherwise, you pay the cost of slicing, copying the span, and doing the relevant bounds checks each iteration.

Reply:
I'm fine with that in this code.  If these helpers are ever moved into MemoryExtensions, we can go to town on optimizing the heck out of them, both with avoiding bounds checks and with adding whatever additional paths are necessary to get the best perf.  For LINQ, I think this is good enough, and there's benefit to no proliferating unsafe code here.